### PR TITLE
lang/funcs: added parseint function

### LINKED ISSUE #22584: tonumber() does not support 0x prefixed hexadecimal strings
### Terraform Version
0.12.7

### Terraform Configuration Files
```hcl
variable "v1" {
    default = "0x99"
}
output "input-v1" {
    value = tonumber(var.v1)
}
```

### Expected Behavior
Hex string is parsed, and converted to a number (decimal 153).

### Actual Behavior
```
Error: Invalid function argument

  on main.tf line 5, in output "input-v1":
   5:     value = tonumber(var.v1)
    |----------------
    | var.v1 is "0x99"

Invalid value for "v" parameter: cannot convert "0x99" to number; given string
must be a decimal representation of a number.
```

### Additional Context

The above example is a reduced case. In the real life use case, the hex string is generated from the substring of a hash to deterministically determine AZ placement without placement changing if other nodes are torn down (i.e. can't use index.count) - so this must be represented as a string and converted to a number for lookup in a list. Worked under 0.11 and previous, so considered a bug in 0.12.

### PR DESCRIPTION
**Pull request by jeet-parekh**
Added a function to parse a string to an integer for a specified base.

Closes #22584. Also see #22585.

### CODE REVIEW

File: website/docs/configuration/functions/parseint.html.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
+---
+layout: "functions"
+page_title: "parseint - Functions - Configuration Language"
+sidebar_current: "docs-funcs-numeric-parseint"
```
Comment (apparentlymart) [inline website/docs/configuration/functions/parseint.html.md]: It's import also to link to this page from the functions navigation bar. You can find the template for that in `website/layouts/functions.erb`.

File: lang/funcs/number.go [hunk 2]
```diff
@@ -128,6 +130,47 @@ var SignumFunc = function.New(&function.Spec{
...
+
+		if err = gocty.FromCtyValue(args[1], &base); err != nil {
+			return cty.UnknownVal(cty.Number), err
+		}
+
+		if base < 2 || base > 32 {
+			return cty.UnknownVal(cty.Number), errors.New("base should be within range 2 <= base <= 32")
+		}
+
+		parsedNum, err := strconv.ParseInt(num, base, 0)
```
[the remark this answers is not in the document]
Comment (jeet-parekh) [reply lang/funcs/number.go]: Makes sense. I'll make the changes.

File: lang/funcs/number.go [hunk 3]
```diff
@@ -128,6 +130,47 @@ var SignumFunc = function.New(&function.Spec{
...
+		if err = gocty.FromCtyValue(args[0], &num); err != nil {
+			return cty.UnknownVal(cty.String), err
+		}
+
+		if err = gocty.FromCtyValue(args[1], &base); err != nil {
+			return cty.UnknownVal(cty.Number), err
+		}
+
+		if base < 2 || base > 32 {
+			return cty.UnknownVal(cty.Number), errors.New("base should be within range 2 <= base <= 32")
```
[the remark this answers is not in the document]
Comment (jeet-parekh) [reply lang/funcs/number.go]: I'll try this out.

File: website/docs/configuration/functions/parseint.html.md (hunk 1 above)
Comment (jeet-parekh) [reply website/docs/configuration/functions/parseint.html.md]: Do I directly make changes to `website/layout/functions.erb`? Or something else?

**Comment by jeet-parekh** [comment]
@apparentlymart, I have set the type of the first argument as string. But when I test these changes in the terraform console, it accepts both a number, and a string as the first argument. That's confusing me a bit. Why is that happening?

File: lang/funcs/number.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (jeet-parekh) [reply lang/funcs/number.go]: `Int.SetString` accepts base up to 62. So I'll set the checks accordingly.

**Comment by apparentlymart** [comment]
Hi @jeet-parekh!

The Terraform language automatically converts numbers to strings using base 10 when the context requires a string.

In this particular case I agree it seems a bit strange to allow it, since it probably indicates a user misunderstanding what this function is for or how it works: this function is operating on the individual characters of a string, so populating it with an automatic conversion would be a strange thing to do. It's possible to override that default conversion behavior like this:

* Change the parameter type of the first parameter to `cty.DynamicPseudoType`, which will tell the language runtime to just accept any type and not do any automatic conversions.
* Change the `Type` field in the function definition to be an inline function (rather than a call to `StaticReturnType`) and write logic in that function to check that `args[0].Type()` is `cty.String`, returning an explicit error if not. The `Type` function should still ultimately return `cty.Number, nil` if that check passes, to indicate that the result type will always be a number.

As I was suggesting in my other comments here, we can return this new error using `function.NewArgErrorf` to allow Terraform to indicate specifically that it's the first argument that is incorrect, doing something like this:

```go
    if !args[0].Type().Equals(cty.String) {
         return cty.Number, function.NewArgErrorf(0, "first argument must be string, not %s", args[0].Type().FriendlyName())
    }
```

I suggest we retain the automatic type conversion behavior for the second argument (by leaving it specified as `cty.Number`) because that argument is just a normal parameter interpreted atomically for its value (rather than broken apart and analyzed, as the first argument) and we can therefore remain consistent with the normal language behaviors for that one.

**Comment by jeet-parekh** [comment]
Hi @apparentlymart,

I changed `parseint` to use `math/big`.

What do you think about another function `inttostr`? It would convert a number to a string for a given base.

**Comment by apparentlymart** [comment]
Hi @jeet-parekh! I'm going to look at these changes now, but I first wanted to address your question:

For the moment I don't think we should add a function for formatting numbers because formatting numbers to bases 2, 8, 10, and 16 is already supported via [the `format` function](https://www.terraform.io/docs/configuration/functions/format.html) and I don't think we have a compelling use-case for formatting to any other bases.

If such a use-case _does_ emerge, I expect we'd prefer to address it via enhancements to `format` rather than via a new function, since that function is the primary way to produce non-default string representations of non-string values and includes related features like zero-padding and rounding.

**Comment by apparentlymart** [comment]
Thanks @jeet-parekh!

I made some small changes to fix one failing test, to expand slightly the documentation, and to add some additional tests I used to try out some different inputs. I'm going to merge this now. Thanks again for working on this, and thanks for working through this feedback with me. :tada:

**Comment by jeet-parekh** [comment]
@apparentlymart, did you make any changes to my code to make the tests pass? What were the changes? Thanks!

**Comment by apparentlymart** [comment]
Hi @jeet-parekh,

The test failure was from [the `TestFunctions` test using a number instead of a string](https://github.com/hashicorp/terraform/compare/93bed4dfdfb6020cd8d0b336ad61eed53d4606ee..b3154e8fc637ed6eeb355ed6b3501d5b890d5650#diff-1adca13f767483c66cff02fcc43d6c6cR549).

**Comment by aajain-crest** [comment]
Hi @apparentlymart,

When is this change planned to be released?

**Pull request merged**